server/src/main/java/org/opensearch/cluster/coordination/LeaderChecker.java

```
@@ -313,6 +313,10 @@ public void handleException(TransportException exp) {
                             logger.debug(new ParameterizedMessage("leader [{}] health check failed", leader), exp);
                             leaderFailed(new NodeHealthCheckFailureException("node [" + leader + "] failed health checks", exp));
                             return;
+                        } else if (exp.getCause() instanceof CoordinationStateRejectedException) {
+                            logger.debug(new ParameterizedMessage("leader [{}] rejected coordination state", leader), exp);
+                            leaderFailed(new CoordinationStateRejectedException("node [" + leader + "] rejected coordination state", exp));
+                            return;
```

It's quite possible this check might succeed on the next liveness check interval, failing fast might turn out to be disruptive?

---

Hi @Bukhtawar , So `CoordinationStateRejectedException` is thrown in only two cases during a LeaderCheck,

- If the primary master has already removed the data node from the cluster yet the data node is sending a LeaderCheck to it.
- If the node that received the LeaderCheck request is no longer the primary master

In both cases, the exception thrown is deterministic, ie, doing retries after 1 second will not change the state of the cluster. Instead, the data node should ideally fail the leader check and kick in its peer finder so it can either reconnect to the existing cluster manager, or discover who the new cluster manager is. 

Please let me know what you think of this.

---

Got it, but since the checks are any ways failing fast without waiting for a timeout we only save a two additional retries at 1s interval each. Is saving a couple of seconds something that you are aiming to cut down?

---

Correct. we save those 2 seconds, and if we update the leader_check_interval, we save more. If you see the [linked issue](https://github.com/opensearch-project/OpenSearch/issues/17155), we are trying to deploy the cluster with an encrypting mesh, and every time network resets (during a envoy mesh deployment), we see disconnections for 3-4 second. This PR brings down the disconnection to < 1s, which allows our customer to retry with some backoff and succeed the calls in most of the cases.

---

Ideally for the use-case you are looking for, I highly recommend you extend `_cluster/decommission/` APIs to support decommissioning a node. This should help with a deterministic cluster behaviour for any planned activities

---

Right, that's the plan for long-term. In our current mesh deployment setup there is no way to send an event that the controller is deploying or about to deploy so the node can be decomm'ed. So, we wanted to mitigate this at an application level. 

Additionally, I think this behaviour makes sense in general for OS, retries should not be done for deterministic exceptions like : `NodeHealthCheckFailed`, `ConnectTransportException` and `CoordinationStateRejectedException` . The retries make sense only in the case when the node is too busy to respond to a leader check, or any transient non-deterministic failures.